// @fhevm/mock-utils/_cjs/fhevm/contracts/FHEVMExecutor.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _FHEVMExecutor_fhevmExecutorReadonlyContract, _FHEVMExecutor_fhevmExecutorContractAddress, _FHEVMExecutor_aclAddress, _FHEVMExecutor_hcuLimitAddress, _FHEVMExecutor_inputVerifierAddress, _FHEVMExecutor_version;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FHEVMExecutor = void 0;
const ethers_1 = require("ethers");
const address_js_1 = require("../../utils/address.js");
const error_js_1 = require("../../utils/error.js");
const string_js_1 = require("../../utils/string.js");
const FhevmContractWrapper_js_1 = require("./FhevmContractWrapper.js");
const FHEVMExecutor_itf_js_1 = require("./interfaces/FHEVMExecutor.itf.js");
class FHEVMExecutor extends FhevmContractWrapper_js_1.FhevmCoprocessorContractWrapper {
    constructor() {
        super("FHEVMExecutor");
        _FHEVMExecutor_fhevmExecutorReadonlyContract.set(this, void 0);
        _FHEVMExecutor_fhevmExecutorContractAddress.set(this, void 0);
        _FHEVMExecutor_aclAddress.set(this, void 0);
        _FHEVMExecutor_hcuLimitAddress.set(this, void 0);
        _FHEVMExecutor_inputVerifierAddress.set(this, void 0);
        _FHEVMExecutor_version.set(this, void 0);
    }
    static async create(runner, fhevmExecutorContractAddress, abi, properties) {
        (0, address_js_1.assertIsAddress)(fhevmExecutorContractAddress, "fhevmExecutorContractAddress");
        const fhevmExecutor = new FHEVMExecutor();
        __classPrivateFieldSet(fhevmExecutor, _FHEVMExecutor_fhevmExecutorContractAddress, fhevmExecutorContractAddress, "f");
        __classPrivateFieldSet(fhevmExecutor, _FHEVMExecutor_fhevmExecutorReadonlyContract, new ethers_1.ethers.Contract(fhevmExecutorContractAddress, abi ?? FHEVMExecutor_itf_js_1.FHEVMExecutorPartialInterface, runner), "f");
        __classPrivateFieldSet(fhevmExecutor, _FHEVMExecutor_aclAddress, properties?.aclAddress, "f");
        __classPrivateFieldSet(fhevmExecutor, _FHEVMExecutor_hcuLimitAddress, properties?.hcuLimitAddress, "f");
        __classPrivateFieldSet(fhevmExecutor, _FHEVMExecutor_inputVerifierAddress, properties?.inputVerifierAddress, "f");
        __classPrivateFieldSet(fhevmExecutor, _FHEVMExecutor_version, properties?.version, "f");
        await fhevmExecutor._initialize();
        return fhevmExecutor;
    }
    get readonlyContract() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorReadonlyContract, "f") !== undefined, `FHEVMExecutor wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorReadonlyContract, "f");
    }
    get interface() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorReadonlyContract, "f") !== undefined, `FHEVMExecutor wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorReadonlyContract, "f").interface;
    }
    get address() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorContractAddress, "f") !== undefined, `FHEVMExecutor wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorContractAddress, "f");
    }
    get version() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_version, "f") !== undefined, `FHEVMExecutor wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _FHEVMExecutor_version, "f");
    }
    get aclAddress() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_aclAddress, "f") !== undefined, `FHEVMExecutor wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _FHEVMExecutor_aclAddress, "f");
    }
    get hcuLimitAddress() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_hcuLimitAddress, "f") !== undefined, `FHEVMExecutor wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _FHEVMExecutor_hcuLimitAddress, "f");
    }
    get inputVerifierAddress() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_inputVerifierAddress, "f") !== undefined, `FHEVMExecutor wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _FHEVMExecutor_inputVerifierAddress, "f");
    }
    async _initialize() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorReadonlyContract, "f") !== undefined, `FHEVMExecutor wrapper is not initialized`);
        if (!__classPrivateFieldGet(this, _FHEVMExecutor_aclAddress, "f")) {
            __classPrivateFieldSet(this, _FHEVMExecutor_aclAddress, await this._callOrThrow(__classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorReadonlyContract, "f").getACLAddress(), "getACLAddress()"), "f");
        }
        (0, address_js_1.assertIsAddress)(__classPrivateFieldGet(this, _FHEVMExecutor_aclAddress, "f"), "aclAddress");
        if (!__classPrivateFieldGet(this, _FHEVMExecutor_hcuLimitAddress, "f")) {
            __classPrivateFieldSet(this, _FHEVMExecutor_hcuLimitAddress, await this._callOrThrow(__classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorReadonlyContract, "f").getHCULimitAddress(), "getHCULimitAddress()"), "f");
        }
        (0, address_js_1.assertIsAddress)(__classPrivateFieldGet(this, _FHEVMExecutor_hcuLimitAddress, "f"), "hcuLimitAddress");
        if (!__classPrivateFieldGet(this, _FHEVMExecutor_inputVerifierAddress, "f")) {
            __classPrivateFieldSet(this, _FHEVMExecutor_inputVerifierAddress, await this._callOrThrow(__classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorReadonlyContract, "f").getInputVerifierAddress(), "getInputVerifierAddress()"), "f");
        }
        (0, address_js_1.assertIsAddress)(__classPrivateFieldGet(this, _FHEVMExecutor_inputVerifierAddress, "f"), "inputVerifierAddress");
        if (!__classPrivateFieldGet(this, _FHEVMExecutor_version, "f")) {
            __classPrivateFieldSet(this, _FHEVMExecutor_version, await this._callOrThrow(__classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorReadonlyContract, "f").getVersion(), "getVersion()"), "f");
        }
        (0, string_js_1.assertIsString)(__classPrivateFieldGet(this, _FHEVMExecutor_version, "f"), "version");
    }
}
exports.FHEVMExecutor = FHEVMExecutor;
_FHEVMExecutor_fhevmExecutorReadonlyContract = new WeakMap(), _FHEVMExecutor_fhevmExecutorContractAddress = new WeakMap(), _FHEVMExecutor_aclAddress = new WeakMap(), _FHEVMExecutor_hcuLimitAddress = new WeakMap(), _FHEVMExecutor_inputVerifierAddress = new WeakMap(), _FHEVMExecutor_version = new WeakMap();
