// @fhevm/mock-utils/_cjs/fhevm/contracts/HCULimit.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _HCULimit_hcuLimitContract, _HCULimit_hcuLimitContractAddress, _HCULimit_fhemExecutorAddress, _HCULimit_version;
Object.defineProperty(exports, "__esModule", { value: true });
exports.HCULimit = void 0;
const ethers_1 = require("ethers");
const address_js_1 = require("../../utils/address.js");
const error_js_1 = require("../../utils/error.js");
const string_js_1 = require("../../utils/string.js");
const FhevmContractWrapper_js_1 = require("./FhevmContractWrapper.js");
const HCULimit_itf_js_1 = require("./interfaces/HCULimit.itf.js");
class HCULimit extends FhevmContractWrapper_js_1.FhevmCoprocessorContractWrapper {
    constructor() {
        super("HCULimit");
        _HCULimit_hcuLimitContract.set(this, void 0);
        _HCULimit_hcuLimitContractAddress.set(this, void 0);
        _HCULimit_fhemExecutorAddress.set(this, void 0);
        _HCULimit_version.set(this, void 0);
    }
    static async create(runner, hcuLimitContractAddress, abi, properties) {
        (0, address_js_1.assertIsAddress)(hcuLimitContractAddress, "hcuLimitContractAddress");
        if (properties !== undefined) {
            throw new error_js_1.FhevmError("Not yet implemented");
        }
        const hcuLimit = new HCULimit();
        __classPrivateFieldSet(hcuLimit, _HCULimit_hcuLimitContractAddress, hcuLimitContractAddress, "f");
        __classPrivateFieldSet(hcuLimit, _HCULimit_hcuLimitContract, new ethers_1.ethers.Contract(hcuLimitContractAddress, abi ?? HCULimit_itf_js_1.HCULimitPartialInterface, runner), "f");
        await hcuLimit._initialize();
        return hcuLimit;
    }
    get readonlyContract() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f") !== undefined, `HCULimit wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f");
    }
    get interface() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f") !== undefined, `HCULimit wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f").interface;
    }
    get address() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _HCULimit_hcuLimitContractAddress, "f") !== undefined, `HCULimit wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _HCULimit_hcuLimitContractAddress, "f");
    }
    get version() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _HCULimit_version, "f") !== undefined, `HCULimit wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _HCULimit_version, "f");
    }
    get fhemExecutorAddress() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _HCULimit_fhemExecutorAddress, "f") !== undefined, `HCULimit wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _HCULimit_fhemExecutorAddress, "f");
    }
    async _initialize() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f") !== undefined, `HCULimit wrapper is not yet initialized`);
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _HCULimit_fhemExecutorAddress, "f") === undefined, `HCULimit wrapper already initialized`);
        __classPrivateFieldSet(this, _HCULimit_fhemExecutorAddress, await __classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f").getFHEVMExecutorAddress(), "f");
        (0, address_js_1.assertIsAddress)(__classPrivateFieldGet(this, _HCULimit_fhemExecutorAddress, "f"), "fhemExecutorAddress");
        __classPrivateFieldSet(this, _HCULimit_version, await __classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f").getVersion(), "f");
        (0, string_js_1.assertIsString)(__classPrivateFieldGet(this, _HCULimit_version, "f"), "version");
    }
}
exports.HCULimit = HCULimit;
_HCULimit_hcuLimitContract = new WeakMap(), _HCULimit_hcuLimitContractAddress = new WeakMap(), _HCULimit_fhemExecutorAddress = new WeakMap(), _HCULimit_version = new WeakMap();
