// @fhevm/mock-utils/_cjs/fhevm/contracts/ZamaFheDecryptionOracle.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContractAddress;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZamaFheDecryptionOracle = void 0;
const ethers_1 = require("ethers");
const address_js_1 = require("../../utils/address.js");
const error_js_1 = require("../../utils/error.js");
const FhevmContractWrapper_js_1 = require("./FhevmContractWrapper.js");
const KMSVerifier_itf_js_1 = require("./interfaces/KMSVerifier.itf.js");
class ZamaFheDecryptionOracle extends FhevmContractWrapper_js_1.FhevmDecryptionOracleContractWrapper {
    constructor() {
        super("DecryptionOracle");
        _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract.set(this, void 0);
        _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContractAddress.set(this, void 0);
    }
    static async create(runner, zamaFheDecryptionOracleContractAddress, abi) {
        (0, address_js_1.assertIsAddress)(zamaFheDecryptionOracleContractAddress, "zamaFheDecryptionOracleContractAddress");
        const decryptionOracle = new ZamaFheDecryptionOracle();
        __classPrivateFieldSet(decryptionOracle, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContractAddress, zamaFheDecryptionOracleContractAddress, "f");
        __classPrivateFieldSet(decryptionOracle, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, new ethers_1.ethers.Contract(zamaFheDecryptionOracleContractAddress, abi ?? KMSVerifier_itf_js_1.KMSVerifierPartialInterface, runner), "f");
        await decryptionOracle._initialize();
        return decryptionOracle;
    }
    get readonlyContract() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, "f") !== undefined, `ZamaFheDecryptionOracle wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, "f");
    }
    get interface() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, "f") !== undefined, `ZamaFheDecryptionOracle wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, "f").interface;
    }
    async _initialize() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, "f") !== undefined, `ZamaFheDecryptionOracle wrapper is not initialized`);
    }
    get address() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContractAddress, "f") !== undefined, `ZamaFheDecryptionOracle wrapper not initialized`);
        return __classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContractAddress, "f");
    }
}
exports.ZamaFheDecryptionOracle = ZamaFheDecryptionOracle;
_ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract = new WeakMap(), _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContractAddress = new WeakMap();
