// @fhevm/mock-utils/_esm/fhevm/contracts/HCULimit.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _HCULimit_hcuLimitContract, _HCULimit_hcuLimitContractAddress, _HCULimit_fhemExecutorAddress, _HCULimit_version;
import { ethers as EthersT } from "ethers";
import { assertIsAddress } from "../../utils/address.js";
import { FhevmError, assertFhevm } from "../../utils/error.js";
import { assertIsString } from "../../utils/string.js";
import { FhevmCoprocessorContractWrapper } from "./FhevmContractWrapper.js";
import { HCULimitPartialInterface } from "./interfaces/HCULimit.itf.js";
export class HCULimit extends FhevmCoprocessorContractWrapper {
    constructor() {
        super("HCULimit");
        _HCULimit_hcuLimitContract.set(this, void 0);
        _HCULimit_hcuLimitContractAddress.set(this, void 0);
        _HCULimit_fhemExecutorAddress.set(this, void 0);
        _HCULimit_version.set(this, void 0);
    }
    static async create(runner, hcuLimitContractAddress, abi, properties) {
        assertIsAddress(hcuLimitContractAddress, "hcuLimitContractAddress");
        if (properties !== undefined) {
            throw new FhevmError("Not yet implemented");
        }
        const hcuLimit = new HCULimit();
        __classPrivateFieldSet(hcuLimit, _HCULimit_hcuLimitContractAddress, hcuLimitContractAddress, "f");
        __classPrivateFieldSet(hcuLimit, _HCULimit_hcuLimitContract, new EthersT.Contract(hcuLimitContractAddress, abi ?? HCULimitPartialInterface, runner), "f");
        await hcuLimit._initialize();
        return hcuLimit;
    }
    get readonlyContract() {
        assertFhevm(__classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f") !== undefined, `HCULimit wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f");
    }
    get interface() {
        assertFhevm(__classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f") !== undefined, `HCULimit wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f").interface;
    }
    get address() {
        assertFhevm(__classPrivateFieldGet(this, _HCULimit_hcuLimitContractAddress, "f") !== undefined, `HCULimit wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _HCULimit_hcuLimitContractAddress, "f");
    }
    get version() {
        assertFhevm(__classPrivateFieldGet(this, _HCULimit_version, "f") !== undefined, `HCULimit wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _HCULimit_version, "f");
    }
    get fhemExecutorAddress() {
        assertFhevm(__classPrivateFieldGet(this, _HCULimit_fhemExecutorAddress, "f") !== undefined, `HCULimit wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _HCULimit_fhemExecutorAddress, "f");
    }
    async _initialize() {
        assertFhevm(__classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f") !== undefined, `HCULimit wrapper is not yet initialized`);
        assertFhevm(__classPrivateFieldGet(this, _HCULimit_fhemExecutorAddress, "f") === undefined, `HCULimit wrapper already initialized`);
        __classPrivateFieldSet(this, _HCULimit_fhemExecutorAddress, await __classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f").getFHEVMExecutorAddress(), "f");
        assertIsAddress(__classPrivateFieldGet(this, _HCULimit_fhemExecutorAddress, "f"), "fhemExecutorAddress");
        __classPrivateFieldSet(this, _HCULimit_version, await __classPrivateFieldGet(this, _HCULimit_hcuLimitContract, "f").getVersion(), "f");
        assertIsString(__classPrivateFieldGet(this, _HCULimit_version, "f"), "version");
    }
}
_HCULimit_hcuLimitContract = new WeakMap(), _HCULimit_hcuLimitContractAddress = new WeakMap(), _HCULimit_fhemExecutorAddress = new WeakMap(), _HCULimit_version = new WeakMap();
