// @fhevm/mock-utils/_esm/fhevm/contracts/ZamaFheDecryptionOracle.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContractAddress;
import { ethers as EthersT } from "ethers";
import { assertIsAddress } from "../../utils/address.js";
import { assertFhevm } from "../../utils/error.js";
import { FhevmDecryptionOracleContractWrapper } from "./FhevmContractWrapper.js";
import { KMSVerifierPartialInterface } from "./interfaces/KMSVerifier.itf.js";
// Shareable
export class ZamaFheDecryptionOracle extends FhevmDecryptionOracleContractWrapper {
    constructor() {
        super("DecryptionOracle");
        _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract.set(this, void 0);
        _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContractAddress.set(this, void 0);
    }
    static async create(runner, zamaFheDecryptionOracleContractAddress, abi) {
        assertIsAddress(zamaFheDecryptionOracleContractAddress, "zamaFheDecryptionOracleContractAddress");
        const decryptionOracle = new ZamaFheDecryptionOracle();
        __classPrivateFieldSet(decryptionOracle, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContractAddress, zamaFheDecryptionOracleContractAddress, "f");
        __classPrivateFieldSet(decryptionOracle, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, new EthersT.Contract(zamaFheDecryptionOracleContractAddress, abi ?? KMSVerifierPartialInterface, runner), "f");
        await decryptionOracle._initialize();
        return decryptionOracle;
    }
    get readonlyContract() {
        assertFhevm(__classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, "f") !== undefined, `ZamaFheDecryptionOracle wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, "f");
    }
    get interface() {
        assertFhevm(__classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, "f") !== undefined, `ZamaFheDecryptionOracle wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, "f").interface;
    }
    async _initialize() {
        assertFhevm(__classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract, "f") !== undefined, `ZamaFheDecryptionOracle wrapper is not initialized`);
    }
    get address() {
        assertFhevm(__classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContractAddress, "f") !== undefined, `ZamaFheDecryptionOracle wrapper not initialized`);
        return __classPrivateFieldGet(this, _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContractAddress, "f");
    }
}
_ZamaFheDecryptionOracle_zamaFheDecryptionOracleContract = new WeakMap(), _ZamaFheDecryptionOracle_zamaFheDecryptionOracleContractAddress = new WeakMap();
