// @fhevm/mock-utils/_cjs/fhevm/decryptionOracle/abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DecryptionOraclePartialInterface = void 0;
const ethers_1 = require("ethers");
exports.DecryptionOraclePartialInterface = new ethers_1.ethers.Interface([
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "uint256",
                name: "counter",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "requestID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "bytes32[]",
                name: "cts",
                type: "bytes32[]",
            },
            {
                indexed: false,
                internalType: "address",
                name: "contractCaller",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bytes4",
                name: "callbackSelector",
                type: "bytes4",
            },
        ],
        name: "DecryptionRequest",
        type: "event",
    },
]);
